[Merged by Bors] - feat(RingTheory/Algebraic): transitivity of algebraicity

alreadydone opened the pull request:
Main results:

* `Subalgebra.algebraicClosure`: If `R` is a domain and `S` is an arbitrary `R`-algebra, then the elements of `S` that are algebraic over `R` form a subalgebra. (The proofs are due to Matthé van der Lee on [MathOverflow](https://mathoverflow.net/questions/482944/when-do-algebraic-elements-form-a-subalgebra#comment1257824_482944).)

* `Algebra.IsAlgebraic.trans'`: If `A/S/R` is a tower of algebras and both `A/S` and `S/R` are algebraic, then `A/R` is also algebraic, provided that `S` has no zero divisors and `algebraMap S A` is injective (so `S` can be regarded as an `R`-subalgebra of `A`).

* `Transcendental.extendScalars`: an element of an `R`-algebra that is transcendental over `R` remains transcendental over any algebraic `R`-subalgebra that has no zero divisors.

* `IsLocalization.isAlgebraic`: a localization of a nontrivial commutative ring is always algebraic. Used to golf `exists_integral_multiples`.

Auxiliary result:

* `IsAlgebraic.iff_exists_smul_integral`: If `R` is reduced and `S` is an `R`-algebra with injective `algebraMap`, then an element of `S` is algebraic over `R` iff some `R`-multiple is integral over `R`.

---

`AlgebraicIndependent.extendScalars` will be PR'd once #19606 is in.

- [x] depends on: #19592 
- [x] depends on: #19665

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

acmepjz commented:
Maybe you want to change the base branch to `master`?

alreadydone commented:
It's easier to review if the changes from #19592 (3 files touched) are excluded.

acmepjz reviewed (commented):
Looks good to me! I haven't spot any other problems.

Do you think it's good to add the relation of `integralClosure`, `Subalgebra.algebraicClosure` and the intermediate field version `algebraicClosure`? For example, sometimes `integralClosure` and `Subalgebra.algebraicClosure` are equal, and some of the results on `IsAlgebraic/Transcendental` and `integralClosure`, `Subalgebra.algebraicClosure` also holds for `algebraicClosure`.

alreadydone commented:
> Do you think it's good to add the relation of `integralClosure`, `Subalgebra.algebraicClosure` and the intermediate field version `algebraicClosure`? For example, sometimes `integralClosure` and `Subalgebra.algebraicClosure` are equal,

I've added `algebraicClosure_toSubalgebra` (the intermediate field version) and `Subalgebra.algebraicClosure_eq_integralClosure` (over a field). Apart from fields they are probably not equal.

> and some of the results on `IsAlgebraic/Transcendental` and `integralClosure`, `Subalgebra.algebraicClosure` also holds for `algebraicClosure`.

I renamed the `Transcendental.algebraicClosure` to `subalgebraAlgebraicClosure` and used the original name for the intermediate field version. I can add `AlgebraicIndependent..algebraicClosure` with the future PR, but there isn't a good home for it apparently ...

alreadydone commented on Mathlib/RingTheory/Algebraic/Integral.lean:
```diff
@@ -146,15 +156,249 @@ end Field
+  rw [Transcendental, not_not] at ha ⊢
+  exact ha.restrictScalars_of_isIntegral _ inj
+
+lemma extendScalars [NoZeroDivisors S] [Algebra.IsAlgebraic R S]
+    (inj : Function.Injective (algebraMap S A)) : Transcendental S a := by
+  contrapose ha
+  rw [Transcendental, not_not] at ha ⊢
+  exact ha.restrictScalars _ inj
```
There is some evidence that the [NoZeroDivisors S] condition might be weakened. If S is a localization of R, then the injectivity of `R[X] → A` (from `Transcendental R a`) implies the injectivity of `S[X] → S ⊗[R] A`, but since A/S/R is a tower and S/R is a localization, `S ⊗[R] A` is just `A`, and we get `Transcendental S a` without requiring S to have no zero divisors. Of course, `Transcendental R a` implies `algebraMap R S` is injective, so the submonoid can't contain any zero divisors.

alreadydone commented on Mathlib/RingTheory/Algebraic/Basic.lean:
```diff
@@ -93,6 +108,17 @@ theorem transcendental_iff_ker_eq_bot {x : A} :
+  isAlgebraic a := isAlgebraic_iff_not_injective.mpr
+    fun inj ↦ h <| by convert inj.comp C_injective; ext; simp
+
+theorem Algebra.injective_of_transcendental (h : Algebra.Transcendental R A) :
```
```suggestion
theorem Algebra.injective_of_transcendental [Algebra.Transcendental R A] :
```

alreadydone commented on Mathlib/FieldTheory/AlgebraicClosure.lean:
```diff
@@ -100,15 +103,18 @@ variable (F E K)
 
 end algebraicClosure
 
+theorem Transcendental.algebraicClosure {a : E} (ha : Transcendental F a) :
```
```suggestion
protected theorem Transcendental.algebraicClosure {a : E} (ha : Transcendental F a) :
```

alreadydone commented on Mathlib/RingTheory/Algebraic/Basic.lean:
```diff
@@ -93,6 +109,17 @@ theorem transcendental_iff_ker_eq_bot {x : A} :
+  isAlgebraic a := isAlgebraic_iff_not_injective.mpr
+    fun inj ↦ h <| by convert inj.comp C_injective; ext; simp
+
+theorem Algebra.injective_of_transcendental [Algebra.Transcendental R A] :
+    Function.Injective (algebraMap R A) := by
```
```suggestion
theorem Algebra.injective_of_transcendental [h : Algebra.Transcendental R A] :
    Function.Injective (algebraMap R A) := by
```

jcommelin reviewed (commented):
Thanks :tada:

bors merge
